src/compute-client/src/controller/instance.rs

```diff
@@ -999,18 +999,17 @@ where
 
         // Validate sources have `since.less_equal(as_of)`.
         for source_id in dataflow.source_imports.keys() {
-            let since = &self
-                .storage_controller
-                .collection(*source_id)
-                .map_err(|_| DataflowCreationError::CollectionMissing(*source_id))?
-                .read_capabilities
-                .frontier();
-            if !(timely::order::PartialOrder::less_equal(since, &as_of.borrow())) {
-                Err(DataflowCreationError::SinceViolation(*source_id))?;
-            }
+            // We cannot assert on the _current_ implied capability/since of
+            // depended-upon storage collections. Their frontiers might move
+            // along concurrently. Our only hope is that our callers put in
+            // place the right read holds, which prevents sinces from advancing.
+            //
+            // Below, where we install read holds using
+            // update_read_capabilities, that call will fail if someone messed
+            // up the read holds.
 
             storage_dependencies.push(*source_id);
-            replica_write_frontier.join_assign(&since.to_owned());
+            replica_write_frontier.join_assign(as_of);
```

Comment:
Hm, this is annoying. We used the `since` here for a reason, see the comment on `replica_write_frontier` above. I fear with this change we introduce the risk of seeing replicas panic with "cannot serve requested as_of" again. Specifically, the following sequence can happen:

* Controller is instructed to install a dataflow with `as_of = t1`.
* Controller installs read holds on the dataflow inputs at `t1`.
* Controller sends a `CreateDataflow(as_of=t1)` command.
* Replica receives the `CreateDataflow(as_of=t1)` command but is delayed in rendering the dataflow.
* envd restarts and performs bootstrapping, which selects `as_of = t2` with `t2 > t1`.
* Controller is instructed to install the same dataflow but with `as_of = t2`.
* Controller installs read holds on the dataflow inputs at `t2` and, notably, not at `t1`.
* Storage allows compaction of the dataflow inputs up to `t2`.
* Replica gets around to rendering the `CreateDataflow(as_of=t1)` and panics when trying to open read handles for its inputs.

Is there a safe way for the compute controller to install a read hold at either `read_capabilities.frontier()` or `implied_capability` here? I think both should be fine because during bootstrap the `implied_capability` should be equal to the since (adapter needs to wait with applying read policies until we had a chance to install read holds for all dataflows). If we could get such a read hold, we could use it to build the initial `replica_write_frontier` and we could also use it to validate the `as_of` (which is why I assume getting the read hold is not possible currently).